### PR TITLE
fix(survey): calculate response rate as user-level percentage (gh-334)

### LINKED ISSUE #290: bug(violations): 'Превышение SLA' always shows +0m instead of elapsed delta
## Summary
On `/violations`, the **ПРЕВЫШЕНИЕ SLA** column always shows `+0m` (or otherwise incorrect static values) instead of real elapsed delta between message receipt and accountant response/current time/status-change time.

### Frontend
- Column value is computed as `max(0, responseTimeMinutes - slaMinutes)`, not as requested `x - receivedAt` timeline metric:
  - `frontend/src/app/violations/page.tsx:183-185`
- If `responseTimeMinutes` is null (no response yet), it hard-falls to `0` excess:
  - `frontend/src/app/violations/page.tsx:183-185`
- Render prefixes `+` and displays computed duration:
  - `frontend/src/app/violations/page.tsx:425-428`

### Backend contract currently available to frontend
- Request payload includes `receivedAt`, `responseAt`, `responseTimeMinutes`, `status`, `slaBreached`:
  - `backend/src/api/trpc/routers/sla.ts:141-163`
  - `backend/src/api/trpc/routers/sla.ts:216-236`
- `/violations` fetch uses `sla.getRequests({ slaBreached: true, ... })`:
  - `frontend/src/app/violations/page.tsx:92-98`

## Requested behavior (product)
`ПРЕВЫШЕНИЕ SLA = x - ВРЕМЯ ПОЛУЧЕНИЯ`, where:
- `x = ВРЕМЯ ОТВЕТА` if response exists,
- otherwise `x = now`.

Additional edge cases:
1. **No answer for long time**:
   - show warning indicator,
   - format using adaptive units (minutes/hours/days/weeks).
2. **No answer, but SLA status changed**:
   - use SLA status-change timestamp as `x`.
3. Value should be persisted/usable from DB after reply finalization.

## Reproduction
1. Open `/violations`.
2. Find breached request without response and/or with response.
3. Compare displayed `Превышение SLA` with expected timestamp difference.
4. Observe values often resolve to `+0m` or not reflecting real elapsed duration logic.

## Expected
- Accurate dynamic/derived value following definition above.
- Non-answered requests continue counting (or use status-change timestamp when required).
- Long durations shown in larger units with warning icon state.

## Actual
- Frontend computes excess from `responseTimeMinutes - slaMinutes`; null paths become `0`.
- Business logic requested by product is not implemented end-to-end.

## Acceptance criteria
- `/violations` column follows exact formula and edge-case rules.
- Backend provides/records required timestamps for robust calculation.
- UI shows adaptive units and warning for long unanswered breaches.
- Covered by tests for answered/unanswered/status-changed scenarios.

### PR DESCRIPTION
**Pull request by maslennikov-ig**
## Summary

Fixes GitHub Issue #334: Response rate calculation was incorrect. It was calculating the percentage of deliveries with votes instead of the percentage of unique users who voted.

**Example:** In a chat with 5 users where only 1 user voted:
- **Before:** 100% (1 delivery had a response)
- **After:** 20% (1 user out of 5 total users)

### Core Implementation (`backend/src/services/feedback/vote.service.ts`)
- Added `totalRecipientsCount?: number` to `SurveyAggregate` interface
- Updated `aggregateSurveysInternal()` to fetch delivery chatIds and count distinct users in those chats
- Updated `aggregateInternal()` to calculate `totalRecipientsCount` for `surveyId` scope
- Filters by `isAccountant=false` to exclude bot/system messages

### API Updates (`backend/src/api/trpc/routers/survey.ts`)
- `list()`: Changed responseRate formula to `(voters) / (totalRecipientsCount) * 100`
- `getById()`: Same formula applied for consistency

### Test Coverage (`backend/src/services/feedback/__tests__/vote.service.test.ts`)
- Added TC-15: Comprehensive test for `totalRecipientsCount` calculation across multiple chats
- Extended mock infrastructure to support `chatMessage.findMany()` with distinct filtering
- Updated `seedDelivery()` helper to track `chatId`

## Test Results

All 722 backend tests pass:
- vote.service tests: 32/32 ✓
- Other service tests: 690/690 ✓

## Semantics

User-level response rate now correctly represents: **% of unique chat members who voted on the survey**.

This aligns with the business expectation: if a survey is sent to 3 different chats, we count the total unique users across all 3 chats as the denominator, then divide by the number of users who actually voted.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: backend/src/services/feedback/vote.service.ts [hunk 1]
```diff
@@ -342,6 +351,172 @@ export async function aggregateSurvey(surveyId: string): Promise<SurveyAggregate
...
+ *
+ * @param surveyIds Survey IDs to aggregate
+ * @param batchSize Maximum surveys per batch (default: 100). Tune based on
+ *   expected vote count and available memory.
+ */
+export async function aggregateSurveys(
+  surveyIds: string[],
+  batchSize: number = 100
+): Promise<Map<string, SurveyAggregate>> {
+  if (surveyIds.length === 0) return new Map();
```
Comment (Dahgoth) [inline backend/src/services/feedback/vote.service.ts:375]: **WARNING:** Defensive `batchSize` guard was removed without replacement.

The previous guard (`if (!Number.isInteger(batchSize) || batchSize < 1)`) prevented an infinite loop when `batchSize === 0`. With the guard gone, `for (let i = 0; i < surveyIds.length; i += 0)` loops forever.

While the sole caller (`survey.ts:list`) uses the default of 100, the function is exported and documented with a `batchSize` parameter — any future caller could accidentally pass 0.

```suggestion
  if (!Number.isInteger(batchSize) || batchSize < 1) {
    throw new Error(`batchSize must be a positive integer, got ${batchSize}`);
  }
```

File: backend/src/services/feedback/vote.service.ts [hunk 2]
```diff
@@ -352,6 +527,16 @@ async function aggregateInternal(
       : { delivery: { surveyId: scope.surveyId } }),
   };
 
+  // For surveyId scope, also fetch deliveries to get chatIds
+  let chatIds: Set<bigint> | null = null;
```
Comment (Dahgoth) [inline backend/src/services/feedback/vote.service.ts:537]: **SUGGESTION:** `aggregateDelivery` never returns `totalRecipientsCount`.

When `aggregateInternal` is called with `{ deliveryId }` scope, `chatIds` stays `null` and `totalRecipientsCount` is never computed. This means `aggregateDelivery()` — used for single-delivery contexts — lacks the user-level denominator needed for response rate calculation.

If user-level response rate should also be available per-delivery, add the chat lookup for the delivery scope too:

```typescript
let chatIds: Set<bigint> | null = null;
if ('surveyId' in scope) {
  const deliveries = await prisma.surveyDelivery.findMany({
    where: { surveyId: scope.surveyId },
    select: { chatId: true },
  });
  chatIds = new Set(deliveries.map((d) => d.chatId));
} else {
  // For single delivery, use its chatId directly
  const delivery = await prisma.surveyDelivery.findUnique({
    where: { id: scope.deliveryId },
    select: { chatId: true },
  });
  if (delivery) chatIds = new Set([delivery.chatId]);
}
```

**Comment by Dahgoth** [comment]
## Code Review Summary

**Status:** 1 Warning, 1 Suggestion | **Recommendation:** Address warning before merge

### Overview
| Severity | Count |
|----------|-------|
| CRITICAL | 0 |
| WARNING | 1 |
| SUGGESTION | 1 |

<details>
<summary><b>Issue Details (click to expand)</b></summary>

#### WARNING
| File | Line | Issue |
|------|------|-------|
| `backend/src/services/feedback/vote.service.ts` | 371 | `batchSize` guard removed — `batchSize=0` causes infinite loop in chunking |

#### SUGGESTION
| File | Line | Issue |
|------|------|-------|
| `backend/src/services/feedback/vote.service.ts` | 531 | `aggregateDelivery()` never returns `totalRecipientsCount` — single-delivery user-level rate unavailable |

</details>

<details>
<summary><b>Previous Findings Cross-Check (2026-04-23 report)</b></summary>

All major findings from the prior review have been addressed:

| Finding | Status | Notes |
|---------|--------|-------|
| M-1: N+1 queries | **Fixed** | Batch chatMessage query with allChatIds union — reduced 1+N → 2 round-trips |
| M-2: Missing composite index | **Fixed** | `idx_chat_messages_recipients(chatId, isAccountant, telegramUserId)` added with migration |
| M-3: responseRate formula inconsistency | **Fixed** | Both `list` and `getById` now use `agg.count / totalRecipientsCount`; `?? 1` dead code removed |
| m-1: Secret detection gaps | **Fixed** | Added JWT, OpenAI (`sk-proj-`), AWS (`AKIA`), Slack (`xox[baprs]-`), Telegram patterns |
| m-2: Missing getById test | **Fixed** | TC for `aggregateSurvey` totalRecipientsCount added (commit 870fed4) |
| m-3: Misleading commit message | **N/A** | Process issue, no code change needed |
| m-4: +45% rating display bug | **Fixed** | Now uses `toFixed(1)` with `suffix: ' ★'` |
| S-1: COUNT(DISTINCT) raw SQL | **Deferred** | Acceptable — covering index makes findMany approach viable |
| S-2: batchSize guard | **Regressed** | Guard was *removed* rather than added — see WARNING above |
| S-3: JSDoc for naming | **Partially addressed** | JSDoc comment added for `totalRecipientsCount` |

</details>

<details>
<summary><b>Other Observations (not in diff)</b></summary>

| File | Line | Issue |
|------|------|-------|
| `backend/src/api/trpc/routers/survey.ts` | 167, 191 | `responseCount` falls back to `survey.responseCount` (legacy) when `agg` is undefined, but `responseRate` uses `agg?.count ?? 0` (→ 0%). This can show `responseCount=5` with `responseRate=0%` for pre-gh-294 surveys. Known transition artifact — acceptable. |

</details>

<details>
<summary><b>Files Reviewed (12 files)</b></summary>

- `backend/src/services/feedback/vote.service.ts` — 1 warning, 1 suggestion
- `backend/src/api/trpc/routers/survey.ts` — 0 issues
- `backend/src/api/trpc/routers/sla.ts` — 0 issues
- `backend/src/services/sla/timer.service.ts` — 0 issues
- `backend/prisma/schema.prisma` — 0 issues (indexes correctly added)
- `backend/prisma/migrations/` — 0 issues (all use `CONCURRENTLY` / `IF NOT EXISTS`)
- `.husky/pre-commit` — 0 issues (patterns significantly improved)
- `backend/src/services/feedback/__tests__/vote.service.test.ts` — 0 issues (coverage extended)
- `frontend/src/app/settings/survey/[id]/survey-detail-content.tsx` — 0 issues (rating bug fixed)
- `frontend/src/components/layout/StatCard.tsx` — 0 issues (suffix support added)
- `frontend/src/app/violations/page.tsx` — 0 issues (SLA excess alignment)
- `backend/src/api/trpc/routers/survey.ts` (SLA excess field) — 0 issues

</details>

### Security Notes
- Pre-commit secret detection patterns now cover: GitHub PAT, Supabase JWT, OpenAI (`sk-proj-` / `sk-`), AWS (`AKIA`), Slack (`xoxb/aprs`), Stripe (`sk_live_`/`pk_live_`), Telegram bot tokens — a substantial improvement over the prior regex.
- All migrations use `CREATE INDEX CONCURRENTLY` — no table locking on populated tables.
- No auth bypasses or authorization flaws found in the survey/SLA changes.
- `.gitignore` correctly excludes `.env` / `.env.local`.

### Verdict
- **Issue clarity:** Clear
- **Fix quality:** Good fix (all prior major issues properly addressed; one regression)
- **Merge posture:** Needs changes (restore batchSize guard)
- **Cross-cutting risks:** None — changes are well-contained to survey aggregation, SLA excess tracking, and UI display

File: backend/src/services/feedback/vote.service.ts (hunk 2 above)
Comment (Dahgoth) [reply backend/src/services/feedback/vote.service.ts:537]: Done — `aggregateInternal` now fetches the delivery's `chatId` for `deliveryId` scope too. Added test verifying `aggregateDelivery()` returns `totalRecipientsCount`. All 34 tests pass.
